daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/DaffodilInputContentHandler.scala

```diff
@@ -0,0 +1,180 @@
+    infosetEvent.prefixMapping = One(NamespaceBinding(pre, uri, infosetEvent.prefixMapping
+      .orNull))
+  }
+
+  override def endPrefixMapping(prefix: String): Unit = {
+    // do nothing
+  }
+
+  override def startElement(uri: String, localName: String, qName: String, atts: Attributes): Unit = {
+    // we need to check if the characters info is all whitespace, if not, it is an error
+    if (characterData.nonEmpty && characterData.toString().trim.nonEmpty) {
+      throw new IllegalContentWhereEventExpected("Non-whitespace characters in complex " +
+        "Element: " + characterData.toString())
+    } else {
+      // reset since it was whitespace only
+      characterData.setLength(0)
+    }
+
+    if (infosetEvent == null) infosetEvent = new DFDL.SaxInfosetEvent
+    else if (!infosetEvent.isEmpty && infosetEvent.localName.isDefined) {
+      // we started another element while we were in the process of building another
+      // this most likely means the first element was complex, either ways, it is ready
+      // for the inputter queue
+      sendToInputter
+    }
+    // use Attributes to determine xsi:nil value
+    val nilIn = atts.getIndex(XMLConstants.W3C_XML_SCHEMA_INSTANCE_NS_URI, "nil")
+    infosetEvent.nilValue = if (nilIn >= 0) {
+      val nilValue = atts.getValue(nilIn)
+      One(nilValue)
+    } else {
+      Nope
+    }
+    // set localName and namespaceURI
+    if (localName.nonEmpty && uri.nonEmpty) {
+      infosetEvent.namespaceURI = One(uri)
+      infosetEvent.localName = One(localName)
+    } else if (qName.nonEmpty) {
+      val qNameArr = qName.split(":")
+
+      if (qNameArr.length > 1) {
+        infosetEvent.localName = qNameArr.lift(1)
+        val qNamePrefix = qNameArr.lift(0).get
+        val in = atts.getIndex(s"xmlns:$qNamePrefix")
+        if (in >= 0) {
+          val attrUri = atts.getValue(in)
+          infosetEvent.namespaceURI = One(attrUri)
+        }
+      } else {
+        infosetEvent.localName = qNameArr.lift(0)
+      }
+    } else {
+      infosetEvent.localName = if(localName.nonEmpty) One(localName) else Nope
+      infosetEvent.namespaceURI = if(uri.nonEmpty) One(uri) else Nope
+    }
+    infosetEvent.eventType = One(StartElement)
+  }
+
+  override def endElement(uri: String, localName: String, qName: String): Unit = {
+    // if infosetEvent is a startElement, send that first
+    if (infosetEvent.eventType.contains(StartElement)) {
+      // any characterData that exists at this point is valid data as padding data has been
+      // taken care of in startElement
+      val maybeNewStr = One(characterData.toString())
+      infosetEvent.simpleText = maybeNewStr
+      characterData.setLength(0)
+      sendToInputter
+    }
+    infosetEvent.eventType = One(EndElement)
+    sendToInputter
+  }
+
+  override def characters(ch: Array[Char], start: Int, length: Int): Unit = {
+    if (infosetEvent == null) {
+      infosetEvent = new DFDL.SaxInfosetEvent
+    }
```

Comment:
Having a single infosetEvent avoid this kind of stuff. We could just always have an infoset event.